Updates to URIs, URLs, and URNs section

mwherman2000 opened the pull request:
Fixes #155 
Fixes #156


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mwherman2000/did-spec/pull/159.html" title="Last updated on Jan 25, 2019, 4:30 PM UTC (9ab5825)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/did-spec/159/dfdac17...mwherman2000:9ab5825.html" title="Last updated on Jan 25, 2019, 4:30 PM UTC (9ab5825)">Diff</a>

davidlehn commented on index.html:
```diff
@@ -259,8 +259,9 @@ <h2>
-DIDs have a foundation in URLs, so it's important to understand how
-the W3C <a href="https://www.w3.org/TR/uri-clarification/">clarified</a> the terms
+Given the potentially confusing relationship between DIDs, URIs, URLs and URNs, 
+so it's important to understand how the
```
```suggestion
it's important to understand how the
```

msporny commented on index.html:
```diff
@@ -286,6 +287,11 @@ <h2>
 changes controllers, etc. URNs are intended to last forever.
           </li>
         </ol>
+        <p>
```
Editorial nit: Would prefer to see this in a bulleted list of some kind since "DIDs are" is repeated unnecessarily.

msporny commented on index.html:
```diff
@@ -259,8 +259,9 @@ <h2>
         <p>
-DIDs have a foundation in URLs, so it's important to understand how
-the W3C <a href="https://www.w3.org/TR/uri-clarification/">clarified</a> the terms
+Given the potentially confusing relationship between DIDs, URIs, URLs and URNs, 
```
Would rather we not start with a negative spin here.

mwherman2000 commented:
> Fine w/ it being merged, note the editorial nits. Please mention if you plan to fix them or not. Fine w/ merge w/o them being fixed, but reserve the right to clean up the text later on (from an editorial perspective).

- Changed the opening sentence to be neutral (inside of negative).
- Removed one occurrence of "DIDs" in the concluding sentence.  Prefer to leave it as a sentence (vs. a bulleted list).
- In `terms.html`, added the following sentence with link back to this section...
> A DID is a <a href="https://w3c-ccg.github.io/did-spec/#uris-urls-and-urns">URL, URI, and a URN</a>.

msporny commented on terms.html:
```diff
@@ -27,7 +27,8 @@
 specification. A specific <a>DID scheme</a> is defined in a
-<a>DID method</a> specification.
+<a>DID method</a> specification. 
+A DID is a <a href="https://w3c-ccg.github.io/did-spec/#uris-urls-and-urns">URL, URI, and a URN</a>. 
```
DIDs are *not* URNs. URNs start with the text "urn:" -- https://tools.ietf.org/html/rfc8141#section-2

mwherman2000 replied:
@msporny Technically, you're correct: DIDs are not URNs according to RFC8141.

...and to top things off, DIDs might not even be URIs according to https://tools.ietf.org/html/rfc3986#section-1.2.2 ...quoting...

> 1.2.2.  Separating Identification from Interaction
> 
>    A common misunderstanding of URIs is that they are only used to refer
>    to accessible resources.  The URI itself only provides
>    identification; access to the resource is neither guaranteed nor
>    implied by the presence of a URI.  Instead, any operation associated
>    with a URI reference is defined by the protocol element, data format
>    attribute, or natural language text in which it appears.
> 
>    Given a URI, a system may attempt to perform a variety of operations
>    on the resource, as might be characterized by words such as "access",
>    "update", "replace", or "find attributes".  Such operations are
>    defined by the protocols that make use of URIs, not by this
>    specification.  However, we do use a few general terms for describing
>    common operations on URIs.  URI "resolution" is the process of
>    determining an access mechanism and the appropriate parameters
>    necessary to dereference a URI; this resolution may require several
>    iterations.  To use that access mechanism to perform an action on the
>    URI's resource is to "dereference" the URI.

This ties in with the comments in: https://github.com/w3c-ccg/did-spec/issues/121

I think we need decide whether a "DID" is just the identifier part (with a separate access protocol spec) so that we can call them URIs and treat them like URIs ...else we in a green field scenario and have to define everything and not reference the URI/URL/URN specs?

msporny replied:
> I think we need decide whether a "DID" (in the context of the draft DID spec) is just the identifier part (with a separate access protocol spec) so that we can call DIDs "URI"s and treat them like URIs

That was decided a while ago (although the spec text most likely does not reflect it accurately yet)...

1. Decentralized Identifiers are just the identifier. 
2. DIDs refer to DID Documents, which are serializations of the Decentralized Identifier Document Data Model, full stop.
3. There are protocols for interacting w/ DIDs (defined in DID Method specifications).

The DID Spec at present, defines 1, 2, and talks about what 3 has to elaborate upon. These things may be split out into separate specifications in time, but at present, they're in the same specification because we didn't want people to have to jump around too much in the early days to see how all of the pieces fit together.

msporny replied:
>  the way the current draft DID spec is written/scoped, might not even be URIs

We should make it clear that there is a URI Definition part of the spec, a data model part of the spec, a serializations part of the spec, and an "everything else" part of the spec.

mwherman2000 replied:
This is headed in the right direction ...how do we get to some point of closure?

msporny replied:
> This is headed in the right direction ...how do we get to some point of closure?

PRs is how we get to closure. Suggest a set of PRs to reorganize the spec into these sections. I expect @rhiaro could probably put something like that together... or I can do it when I get a next pass at the spec (which is going to be end of Feb based on all of the other fires that are being dealt with at present). These are editorial changes, so I expect they can go in w/o much debate.

Once there is a PR in place, we debate it in the PR, and then pull it in the spec. Sometimes the Editors just make the change if they think it's purely editorial and won't cause community objections. This feels like it falls into that category. So, a couple of ways forward... in this order of preference:

1. @rhiaro writes the PR.
2. @mwherman2000 writes the PR.
3. I write the PR.

mwherman2000 replied:
I'm going to have to bow out of this PR.  I've been banging on the DID spec and the INDY ARM for over 2 months now (self-funded). I'm trying to focus on the next layers of the Indy Platform e.g. the SDK (https://hyperonomy.com/2019/01/24/indy-sdk-getting_started-py-visualizations-light-table/).
I can still provide feedback ...very passionate about making this spec valuable for developers.

msporny replied:
> I'm going to have to bow out of this PR.

No problem, you've given us enough of a direction to head in where we can convert this PR into something that'll make it into the spec. Thank you @mwherman2000 ... @rhiaro and I will discuss how we can get your input integrated into the spec over the next couple of weeks.

msporny commented on index.html:
```diff
@@ -286,6 +286,10 @@ <h2>
           </li>
         </ol>
+        <p>
+DIDs are valid URIs; more specifically, they are also valid URLs and URNs. 
```
DIDs are *not* URNs. URNs start with the text "urn:" -- https://tools.ietf.org/html/rfc8141#section-2

mwherman2000 commented:
@dmitrizagidulin There's technical details/differences that matter.  For example, https://github.com/w3c-ccg/did-spec/issues/97

dmitrizagidulin commented:
@mwherman2000 
> There's technical details/differences that matter.

I'm not seeing it.. Issue #97 has nothing to do with whether DIDs are URLs or URNs, it has to do with JSON-LD parsing and serialization mechanics.

But anyways, hopefully the issue has been settled? That DIDs are emphatically _not_ URNs, they do not start with `urn:`, and do not match the URN RFC?

mwherman2000 commented:
> @mwherman2000
> 
> > There's technical details/differences that matter.
> 
> I'm not seeing it.. Issue #97 has nothing to do with whether DIDs are URLs or URNs, it has to do with JSON-LD parsing and serialization mechanics.
> 
> But anyways, hopefully the issue has been settled? That DIDs are emphatically _not_ URNs, they do not start with `urn:`, and do not match the URN RFC?

DIDs might not even be URIs ... see https://github.com/w3c-ccg/did-spec/pull/159#discussion_r251186233

BigBlueHat commented:
I'd avoid the [WHATWG URL spec](https://url.spec.whatwg.org/#urls). It's intended for browser/JavaScript parsing of values in `src` and `href`'s. It's not a general foundation for identification like [RFC 3986](https://tools.ietf.org/html/rfc3986) (and related RFCs). I'd avoid it, unless you're building an HTML parser or JS runtime.

rhiaro commented:
Spec restructuring and terminology clarification is progressing, and I think the various discussions in this thread are resolved enough to close this PR now.
